design/external-authorization-http-support-design.md

```diff
@@ -0,0 +1,165 @@
+# Global HTTP External Authorization
+
+Status: Draft
+
+## Abstract
+
+HTTP Support for External Authorization.
+
+## Background
+See [#4954](https://github.com/projectcontour/contour/issues/4954) for context.
+
+## Goals
+- Support Envoy's L7 External Authorization filter for HTTP Virtual Hosts.
+- Extend the functionality of existing External Authorization.
+
+## Non Goals
+- Supporting more than one External Authorization Configuration for HTTP endpoints.
+
+## High-Level Design
+
+Contour will add HTTP support for Envoy's External Authorozation.
+
+Currently, Contour supports External Authorization only over HTTPS. External Authorization allows for requests to be authenticated before they are forwarded upstream. However, External authorization over HTTPS is incompatible with setups that terminate TLS at the cloud provider's load balancer layer because the Envoy HCM relies on the SNI to choose the correct filter chain. 
+
+A new type, `GlobalHTTPExtAuth`, will be defined as part of the `ContourConfiguration`.
+A `GlobalHTTPExtAuth` config would define a global external authorization configuration for all the HTTP routes. 
+
+## Detailed Design
+
+### Contour Configuration changes
+By using global external authorization, an external authorization service can be configured in the Contour config file.
+This External authorization configuration will be used for all HTTP routes. 
+
+```yaml
+globalHTTPExtAuth:
+  # extensionService is the <namespace>/<name>
+  # of the ExtensionService we created in the
+  # previous step.
+  extensionService: projectcontour-auth/htpasswd
+  # failOpen is whether to allow requests through
+  # if there's an error.
+  failOpen: false
+  # Context is a set of key/value pairs that are sent to the authentication server in the check request.
+  context:
+    header1: value1
+    header2: value2
+    # ResponseTimeout configures maximum time to wait for a check response from the authorization server
+  responseTimeout: 1s
+
+```
+
+### HTTPS Override
+If an upstream has HTTPS enabled, the HTTPS configuration will take precedence. 
+i.e., if global HTTP external authorization is enabled and an upstream is configured with HTTPS without external authorization. Then that particular upstream would not use envoy's external authorization filter. 
+
+## Alternatives Considered
+
+### Envoy Filter Matching API 
+Having a separate external authorization configuration for each upstream is more desirable. We  can achieve this by leveraging Envoy's [Matching API](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/matching/matching_api)
+Note: This feature is an Alpha release, and there are no guarantees that future changes will maintain backward compatibility. 
```

Comment:
Agree that this is probably the more desirable functionality, especially since it is consistent for all types of virtual hosts. Unfortunately I don't think we'll be comfortable making use of it until the matching API becomes stable.